src/index.ts

```diff
@@ -1,51 +1,68 @@
 import { compactDecrypt, importJWK, importPKCS8 } from 'jose'
-import {
-  Client,
-  ClientAuthMethod,
-  generators,
-  Issuer,
-  ResponseType,
-} from 'openid-client'
+import { Client, generators, Issuer } from 'openid-client'
 
+import {
+  API_VERSION,
+  DEFAULT_SCOPE,
+  DEFAULT_SGID_CODE_CHALLENGE_METHOD,
+  SGID_AUTH_METHOD,
+  SGID_SIGNING_ALG,
+  SGID_SUPPORTED_FLOWS,
+} from './constants'
 import { convertPkcs1ToPkcs8 } from './util'
 
-const SGID_SIGNING_ALG = 'RS256'
-const SGID_SUPPORTED_FLOWS: ResponseType[] = ['code']
-const SGID_AUTH_METHOD: ClientAuthMethod = 'client_secret_post'
+type ConstructorParams = {
+  clientId: string
+  clientSecret: string
+  privateKey: string
+  redirectUris?: string[]
+  hostname?: string
+}
+
+type AuthorizationUrlParams = {
+  state: string
+  scope?: string | string[]
+  nonce?: string | null
+  redirectUri?: string
+  codeChallenge: string
+}
+
+type AuthorizationUrlReturn = { url: string; nonce?: string }
+
+type CallbackParams = {
+  code: string
+  nonce?: string | null
+  redirectUri?: string
+  codeVerifier: string
+}
 
 export class SgidClient {
+  private apiVersion: number
   private privateKey: string
-
   private sgID: Client
 
   constructor({
     clientId,
     clientSecret,
     privateKey,
-    redirectUri,
+    redirectUris,
     hostname = 'https://api.id.gov.sg',
-    apiVersion = 1,
-  }: {
-    clientId: string
-    clientSecret: string
-    privateKey: string
-    redirectUri?: string
-    hostname?: string
-    apiVersion?: number
-  }) {
+  }: ConstructorParams) {
+    this.apiVersion = API_VERSION
+
     // TODO: Discover sgID issuer metadata via .well-known endpoint
     const { Client } = new Issuer({
       issuer: new URL(hostname).origin,
-      authorization_endpoint: `${hostname}/v${apiVersion}/oauth/authorize`,
-      token_endpoint: `${hostname}/v${apiVersion}/oauth/token`,
-      userinfo_endpoint: `${hostname}/v${apiVersion}/oauth/userinfo`,
+      authorization_endpoint: `${hostname}/v${this.apiVersion}/oauth/authorize`,
+      token_endpoint: `${hostname}/v${this.apiVersion}/oauth/token`,
+      userinfo_endpoint: `${hostname}/v${this.apiVersion}/oauth/userinfo`,
       jwks_uri: `${new URL(hostname).origin}/.well-known/jwks.json`,
     })
 
     this.sgID = new Client({
       client_id: clientId,
       client_secret: clientSecret,
-      redirect_uris: redirectUri ? [redirectUri] : undefined,
+      redirect_uris: redirectUris,
```

Comment:
hmm i do agree with you that it doesn't really 'add' anything per se.

i went to dig into panva's `openid-client` npm library that we're using under the hood because his `Client` constructor also takes in multiple `redirect_uris`. from what i understand [here](https://github.com/panva/node-openid-client/blob/d571ceb2e74cb703467cf717b4582627f2dcc1af/lib/helpers/client.js#L148), it seems like if 1 URI was provided during construction and NO `redirect_uri` was specified while calling `authorizationUrl`, it will default to that 1 URI. otherwise if `redirect_uris.length !== 1`, it will use whatever `redirect_uri` was provided while calling the respective function without doing any form of validation.

i guess logically it makes sense to not change `redirect_uri` to `redirect_uris` but to stay consistent with other popular libraries changing it might make sense

Reply:
okay after discussing with @PrawiraGenestonlia we've decided to leave the current implementation as is because there is no upside to allowing RPs to register multiple `redirectUris` and might instead cause more confusion.

i will update the comment on the `authorizationUrl` function to describe how the `redirectUri` passed in will override the one passed in during the constructor